### PR TITLE
[BugFix] release tracked fake tensors

### LINKED ISSUE #190093: test_interpolate_propagate_real_tensors_cuda fails CUDA mem-leak check: real CUDA tensor from data-dependent .tolist() retained past dynamo.reset()
### Summary
`test/dynamo/test_misc.py::MiscTestsDeviceCUDA::test_interpolate_propagate_real_tensors_cuda` fails the CUDA memory-leak check (`PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1`, as run in slow-gradcheck): the caching allocator goes from 0 to 512 bytes after the test. Reproduced locally on an H100 on current `main`, deterministic (3/3 runs).

### What leaks, and the minimal trigger
The leak is NOT related to `interpolate`. By varying the repro I confirmed the leak requires exactly three things together -- `torch.compile`, `fake_tensor_propagate_real_tensors=True`, and a data-dependent `.tolist()`:

| variant | leaked bytes |
|---|---|
| full test | 512 |
| `propagate_real_tensors=False` | 0 |
| eager (no `torch.compile`) | 0 |
| static sizes (no `.tolist()`) | 0 |
| `.tolist()` but no `interpolate` (`mask.sum()+h+w`) | 512 |

Repro (leaks 512 B; interpolate removed):
```python
import gc, torch, torch._functorch.config
torch.cuda.init()
gc.collect(); torch.cuda.synchronize(); torch.cuda.empty_cache()
base = torch.cuda.memory_allocated()
with torch._functorch.config.patch(fake_tensor_propagate_real_tensors=True):
    @torch.compile(backend="eager", fullgraph=True)
    def f(mask, box):
        mask = torch.randn(1, 1, 30, 30, device="cuda")
        h, w = box.tolist()
        return mask.sum() + h + w
    f(torch.tensor([30, 30], device="cuda"), torch.tensor([68, 32], device="cuda"))
torch._dynamo.reset(); gc.collect(); torch.cuda.synchronize(); torch.cuda.empty_cache()
print(torch.cuda.memory_allocated() - base)   # 512, expected 0
```

### Retained reference (from gc.get_referrers)
The leaked 512-byte allocation is the **real input CUDA tensor** (`[68, 32]`). It is held by `FakeTensor.real_tensor`, and that `FakeTensor` is held by a `TrackedFake`. The allocation persists after `torch._dynamo.reset()` + `gc.collect()` + `empty_cache()`.

### Verified age
Reproduced back to 2024-06-17: built #187273 and its parent locally -- both leak 512 B. (I did not test earlier commits.)

### Severity
512 bytes per occurrence, only under the opt-in `fake_tensor_propagate_real_tensors` mode with a data-dependent op, and only observable under the CUDA mem-leak checker.

### Interim CI mitigation
`@skipCUDAMemoryLeakCheckIf(True)` on the test exempts it from the leak check while keeping its functional coverage.

Reproduced locally on H100, deterministic.

_Issue filed with the assistance of an AI agent._

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @chauhang @penguinwu @eellison @aorenste @ezyang @bobrenjc93 @aditvenk @laithsakka @bdhirsh

### PR DESCRIPTION
**Pull request by gtnv**
ShapeEnv.tracked_fakes can hold CUDA tensors after compile. this clears it and removes the leak-check skip from the currenet regression

Testing:
- PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 python test/dynamo/test_misc.py MiscTestsDeviceCUDA.test_interpolate_propagate_real_tensors_cuda
  - passed

Closes #190093

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by gtnv** [comment]
@jansel ptal. 

<details><summary>afaik cluade is wrong about these things:</summary>
<p>

> Custom backends keep leaking.

no? bc only the backends with compile state would.

> hold a `weakref` to the real input tensor

its not looking at the real input thensor, its looking at the copy from propagate-real config flag.

> export's post-compile constraint solving reads `dim_constraints`, not `tracked_fakes`.

functional export builds guards first so the gate can stay.

</p>
</details>

**Review by jansel — APPROVED** [review]
